src/atomic/atomic_cell.rs

```diff
@@ -585,19 +585,97 @@ fn can_transmute<A, B>() -> bool {
     mem::size_of::<A>() == mem::size_of::<B>() && mem::align_of::<A>() >= mem::align_of::<B>()
 }
 
-/// Automatically releases a lock when dropped.
-struct LockGuard {
-    lock: &'static AtomicBool,
+/// A simple stamped lock.
+struct Lock {
+    /// The current state of the lock.
+    ///
+    /// All bits except the least significant one hold the current stamp. When locked, the state
+    /// equals 1 and doesn't contain a valid stamp.
+    state: AtomicUsize,
 }
 
-impl Drop for LockGuard {
+impl Lock {
+    /// If not locked, returns the current stamp.
+    ///
+    /// This method should be called before optimistic reads.
+    #[inline]
+    fn optimistic_read(&self) -> Option<usize> {
+        let state = self.state.load(Ordering::Acquire);
+        if state == 1 {
+            None
+        } else {
+            Some(state)
+        }
+    }
+
+    /// Returns `true` if the current stamp is equal to `stamp`.
+    ///
+    /// This method should be called after optimistic reads to check whether they are valid. The
+    /// argument `stamp` should correspond to the one returned by method `optimistic_read`.
+    #[inline]
+    fn validate_read(&self, stamp: usize) -> bool {
+        atomic::fence(Ordering::Acquire);
+        self.state.load(Ordering::SeqCst) == stamp
```

Comment:
The `SeqCst` is here just to make the operation `SeqCst` like the atomic one without locks.

Reply:
I'm not exactly sure if that's enough to make the operation equivalent to `SeqCst`.

cc @jeehoonkang

Reply:
My reasoning is:

If this `SeqCst` load returns a version that is not locked and is equal to the version before reading, then we've read a value that is still the same as the value present at the exact moment this `SeqCst` load happened. So this load operation is when the `AtomicCell::load` really "atomically" happens.

But yeah, I wonder what @jeehoonkang thinks, too. :)

Reply:
Sequence lock (stamped lock) is... strange.  I think the best introduction to the seqlock and its strangeness is [Hans Boehm's slide](http://safari.ece.cmu.edu/MSPC2012/slides_posters/boehm-slides.pdf).  I think we should implement the "version 3" in this slide, as (1) it's correct, as far as Hans (and I) think, and (2) it's the most efficient scheme among those implementable in C/C++.

According to the slide, (1) this load can be `Relaxed`, and (2) the write-lock should issue a `release` fence.  (2) is necessary even if this load is `SeqCst`: otherwise, bad reorderings may happen on the writer side.

Reply:
@Amanieu 

- I think this PR currently doesn't issue a release fence in [fn write()](https://github.com/crossbeam-rs/crossbeam-utils/pull/39/files/2eb9ede6b0ff30779abdbb5fc428fcb352cf0722#diff-d9dcaf34efc6c069d2f10534c4dc480dR623).

- I tend to think `AtomicCell` should have release/acquire semantics rather than sequential consistency.  For this reason, I'd like to propose to lower `SeqCst` to `Release`/`Acquire` for lock-free cases, and I think it's okay to use seqlock for blocking cases.

Reply:
@Amanieu 

> I think that simply adding a SeqCst load at the end is insufficient to guarantee SeqCst semantics for the entire read operation, but I'm not sure.

Oh, I see now. Maybe what's missing is `SeqCst` ordering on the `swap` operation when locking and `SeqCst` on the `store` operation when unlocking. This wouldn't affect generated assembly on x86-64, while on ARM it would turn one `ldxr` instruction into `ldaxr`.

How that affects benchmarks on my Raspberry Pi (ARM):

```
 name                    before ns/iter  after ns/iter  diff ns/iter  diff %  speedup 
 compare_and_swap_u8     36              42                        6  16.67%   x 0.86 
 compare_and_swap_usize  16              16                        0   0.00%   x 1.00 
 concurrent_load_u8      18,394,425      18,495,399          100,974   0.55%   x 0.99 
 concurrent_load_usize   6,796,930       6,797,764               834   0.01%   x 1.00 
 fetch_add_u8            27              32                        5  18.52%   x 0.84 
 fetch_add_usize         19              19                        0   0.00%   x 1.00 
 load_u8                 26              25                       -1  -3.85%   x 1.04 
 load_usize              6               6                         0   0.00%   x 1.00 
 store_u8                26              29                        3  11.54%   x 0.90 
 store_usize             4               4                         0   0.00%   x 1.00 
```

Reply:
@jeehoonkang 

The idea behind `AtomicCell` is to make an atomic primitive that can be explained in simple words without bringing up jargon like `AcqRel` and `SeqCst`.

We've had some discussion in the previous PR on what guarantees we should make in the documentation. I've been thinking a little about this and came up with a simple explanation of what `AtomicCell<T>` is. Here it goes:

* `let a = AtomicCell::new(x)` is like `let m = Mutex::new(Cell::new(x))`.
* `a.load()` is like `m.lock().get()`.
* `a.store(x)` is like `m.lock().set(x)`.
* `a.swap(x)` is like `m.lock().replace(x)`.
* And so on.

So `AtomicCell<T>` is basically `Mutex<Cell<T>>`, except faster. That's it.

I don't think there should be any mention of `AcqRel` or `SeqCst` in the documentation. If you do care about memory orderings and understand them, then `atomic::Atomic<T>` and `std::sync::atomic::Atomic*` allow configuring orderings and are a better primitive for you anyway.

Also, note that the documentation for things like `Arc`, `Mutex`, `RwLock` `Barrier`, `mpsc`, etc. doesn't mention memory orderings either. In my view, `AtomicCell` belongs into the same category of atomic primitives.

Here's a classic example showing the difference between `SeqCst` and `AcqRel`:

```rust
let x = AtomicCell::new(false);
let y = AtomicCell::new(false);
let z = AtomicCell::new(0usize);

scope(|s| {
    s.spawn(|| x.store(true));
    s.spawn(|| y.store(true));

    s.spawn(|| {
        while !x.load() {}
        if y.load() { z.fetch_add(1); }
    });

    s.spawn(|| {
        while !y.load() {}
        if x.load() { z.fetch_add(1); }
    });
}).unwrap();

assert!(z.load() > 0);
```

The assert passes under `SeqCst` semantics, but fails under `AcqRel`. The assert also passes if one replaces `AtomicCell<T>` with `Mutex<Cell<T>>`. Therefore, we should keep the behavior of `AtomicCell<T>` the same.

Reply:
@jeehoonkang The release fence is in the `Drop` impl of `WriteGuard`.

Reply:
@stjepang 

> So `AtomicCell<T>` is basically `Mutex<Cell<T>>`, except faster. That's it.

But unfortunately, the lock-free implementation with `load(SeqCst)` and `store(SeqCst)` is strictly weaker than `Mutex<Cell<T>>`.  The high-level explanation is that two consecutive `load(SeqCst)` are not strictly ordered, while `Mutex<Cell<T>>` guarantees two consecutive reads are strictly ordered.

In general, I agree `Arc` and `Mutex` completely removes concurrency, leaving no places for atomic orderings.  But `RwLock` (e.g. seqlock) and `mpsc` are not such a case, because the usual efficient implementation of those is strictly weaker than `Arc<Mutex<T>>` version of them.

@Amanieu The release fence should be issued at the end of `write_lock()`, not inside `write_unlock()`.

Reply:
@jeehoonkang 

> The high-level explanation is that two consecutive load(SeqCst) are not strictly ordered, while Mutex<Cell<T>> guarantees two consecutive reads are strictly ordered.

Just to make sure I understood, can you also answer whether the following statements are true or false?

1. Two consecutive `load(SeqCst)` are strictly ordered by the C++ standard.
2. Two consecutive `load(SeqCst)` are strictly ordered on x86-64.
3. Two consecutive `load(SeqCst)` are not strictly ordered on ARM.

> The release fence should be issued at the end of write_lock(), not inside write_unlock().

So we just need to change the ordering of the `swap` operation inside `write()` from `Acquire` to `AcqRel`?

Reply:
Oh, maybe what I said was not clear.  Sorry.  By "consecutive loads" I mean ordered by (extended) coherence, not by intra-thread program order.  Suppose two threads A and B are reading the same value from an `AtomicCell<T>`.  If it's lock-free, two reads are not ordered at all in C++, ARMv8, and ARMv7.  (In x86, probably the order doesn't matter at the beginning.)  But if it's based on a spinlock, then two loads are strictly ordered, either A's read before B's read or the other way around.

Regarding seqlock, acquire and release **fence** should be issued.  AFAIK, in "version 3", `W` issues a release fence right after updating the stamp value in `write_lock()`, and `R` issues an acquire fence just before reading the stamp value in the validation phase.  If `R` reads any value in the protected data written by `W`, then there's a acquire/release synchronization between those fences, forcing `R` to read the stamp value updated by `W` or later

Reply:
> If it's lock-free, two reads are not ordered at all in C++, ARMv8, and ARMv7.

Doesn't at least the C++ standard guarantee that two `SeqCst` reads are always ordered (sequential consistency)?

> W issues a release fence right after updating the stamp value in write_lock()

Did you mean "right before"? (so that writes happen before the stamp value is updated)

> If R reads any value in the protected data written by W, then there's a acquire/release synchronization between those fences

Why not make the `swap` operation by `W` use `AcqRel` so that this operation establishes acquire/release synchronization with the fence issued by `R`? That might be slightly faster than issuing a fence by `W`. Or is that not going to work?

Reply:
- `SeqCst` loads and stores have very vague meaning in C++.  As a theorem (not normative), C++ guarantees that **all** atomic accesses are `SeqCst` and there's no data race for non-atomic locations in a program, then the program's semantics is sequentially consistent (so-called "DRF1" or "DRF-SC"; here "DRF" stands for data-race freedom).  But when even a single access to an irrelevant location is e.g. `Relaxed`, then the program's semantics may no longer be sequentially consistent.

  In short, the semantics of `SeqCst loads and stores is surprisingly weaker than expected.  In particular, it's strictly weaker than spinlock-protected accesses.  For the latest discussion on this subject, [the RC11 paper](https://plv.mpi-sws.org/scfix/) is worth reading.

- I meant "right after", not "right before".  Also, release swap will not work.  (Yes, seqlock is strange...)  The synchronization occurs via data variables, not stamp variable.  Writer issues release fence and then writes a value to data, and if reader reads the new value from data and then issues acquire fence, then after that the reader should read the new stamp value written by the writer (or a later one).

Reply:
> But when even a single access to an irrelevant location is e.g. Relaxed, then the program's semantics may no longer be sequentially consistent.

Whoa, that's really surprising!

By the way, in the previous pull request you said:

>  I think SeqCst obscures the specification: what do we expect from the methods of AtomicCell? I think the word "sequentially consistent" doesn't really help, because practically none understand the term precisely. Even worse, using SeqCst may guarantee some ordering properties which users will rely on, after which we're stuck and cannot optimize it further.

I'd just like to point out that someone might rely on the fact that spinlock-protected `AtomicCell` has sequentially consistent semantics, but then we might switch from spinlock to e.g. `AtomicU16` (once it gets into stable Rust) and thus relax the semantics behind the scenes. So that's a similar problem.

Anyway... I pushed some changes to this PR:

* The load in the validation step now uses `Relaxed` rather than `SeqCst`.
* Added a `Release` fence in `write`.

Does the code look good now?